lib/srv/db/server.go

```diff
@@ -0,0 +1,420 @@
+		closeFunc:     cancel,
+		dynamicLabels: make(map[string]*labels.Dynamic),
+		heartbeats:    make(map[string]*srv.Heartbeat),
+		rdsCACerts:    make(map[string][]byte),
+		middleware: &auth.Middleware{
+			AccessPoint:   config.AccessPoint,
+			AcceptedUsage: []string{teleport.UsageDatabaseOnly},
+		},
+	}
+
+	// Update TLS config to require client certificate.
+	server.TLSConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	server.TLSConfig.GetConfigForClient = getConfigForClient(
+		server.TLSConfig, server.AccessPoint, server.Entry)
+
+	// Perform various initialization actions on each proxied database, like
+	// starting up dynamic labels and loading root certs for RDS dbs.
+	for _, db := range server.Servers {
+		if err := server.initDatabaseServer(localCtx, db); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return server, nil
+}
+
+func (s *Server) initDatabaseServer(ctx context.Context, server services.DatabaseServer) error {
+	if err := s.initDynamicLabels(ctx, server); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := s.initHeartbeat(ctx, server); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := s.initRDSRootCert(ctx, server); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (s *Server) initDynamicLabels(ctx context.Context, server services.DatabaseServer) error {
+	if len(server.GetDynamicLabels()) == 0 {
+		return nil // Nothing to do.
+	}
+	dynamic, err := labels.NewDynamic(ctx, &labels.DynamicConfig{
+		Labels: server.GetDynamicLabels(),
+		Log:    s.Entry,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	dynamic.Sync()
+	s.dynamicLabels[server.GetName()] = dynamic
+	return nil
+}
+
+func (s *Server) initHeartbeat(ctx context.Context, server services.DatabaseServer) error {
+	heartbeat, err := srv.NewHeartbeat(srv.HeartbeatConfig{
+		Context:         s.closeContext,
+		Component:       teleport.ComponentDatabase,
+		Mode:            srv.HeartbeatModeDB,
+		Announcer:       s.AccessPoint,
+		GetServerInfo:   s.getServerInfoFunc(server),
+		KeepAlivePeriod: defaults.ServerKeepAliveTTL,
+		AnnouncePeriod:  defaults.ServerAnnounceTTL/2 + utils.RandomDuration(defaults.ServerAnnounceTTL/10),
+		CheckPeriod:     defaults.HeartbeatCheckPeriod,
+		ServerTTL:       defaults.ServerAnnounceTTL,
+		OnHeartbeat:     s.OnHeartbeat,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.heartbeats[server.GetName()] = heartbeat
+	return nil
+}
+
+func (s *Server) getServerInfoFunc(server services.DatabaseServer) func() (services.Resource, error) {
+	return func() (services.Resource, error) {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		// Update dynamic labels.
+		labels, ok := s.dynamicLabels[server.GetName()]
+		if ok {
+			server.SetDynamicLabels(labels.Get())
+		}
+		// Update CA rotation state.
+		rotation, err := s.GetRotation(teleport.RoleDatabase)
+		if err != nil && !trace.IsNotFound(err) {
+			s.WithError(err).Warn("Failed to get rotation state.")
+		} else {
+			server.SetRotation(*rotation)
```

Comment:
if `trace.IsNotFound(err)`, will rotation be nil or some empty value?
maybe gate this with `if rotation != nil`, just in case